snp_motion_planning/src/cartesian_time_parameterization.hpp

```diff
@@ -0,0 +1,279 @@
+          Eigen::ArrayXd capacity =
+              100.0 * joint_acc.array().abs() / motion_group->getLimits().acceleration_limits.array();
+          std::stringstream ss;
+          ss << "Joint acceleration limit violations at waypoint " << i << ": "
+             << capacity.transpose().format(Eigen::IOFormat(4, 0, " ", "\n", "[", "]")) << " (%% capacity)";
+          throw std::runtime_error(ss.str());
+        }
+
+        // Update the trajectory container
+        trajectory.setData(i, joint_vel, joint_acc, t);
+      }
+    }
+    catch (KDL::Error& e)
+    {
+      std::stringstream ss;
+      ss << "KDL Error #" << e.GetType() << ": " << e.Description();
+      CONSOLE_BRIDGE_logError(ss.str().c_str());
+      return false;
+    }
+    catch (const std::exception& ex)
+    {
+      CONSOLE_BRIDGE_logError(ex.what());
+      return false;
+    }
+
+    return true;
+  }
+
+private:
+  Eigen::VectorXd computeJointVelocity(const KDL::Twist& x_dot, const Eigen::VectorXd& q) const
+  {
+    Eigen::MatrixXd jac = motion_group->calcJacobian(q, tcp);
+
+    Eigen::Index dof = q.size();
+    Eigen::VectorXd q_dot(dof);
+    if (dof == 6)
+    {
+      q_dot = jac.colPivHouseholderQr().solve(fromKDL(x_dot));
+    }
+    else
+    {
+      if (!tesseract_kinematics::solvePInv(jac, fromKDL(x_dot), q_dot))
+        throw std::runtime_error("Failed to solve pseudo-inverse for joint velocity calculation");
+    }
+
+    return q_dot;
+  }
+
+  Eigen::VectorXd computeJointAcceleration(const KDL::Twist& x_dot_dot, const Eigen::VectorXd& q,
+                                           const Eigen::VectorXd& q_dot) const
+  {
+    // Create a Jacobian derivative solver
+    KDL::ChainJntToJacDotSolver solver(kdl_chain_);
+
+    // Compute the derivative of the Jacobian
+    KDL::JntArrayVel jnt_array;
+    jnt_array.q.data = q;
+    jnt_array.qdot.data = q_dot;
+    KDL::Twist jac_dot_q_dot;
+    {
+      int error = solver.JntToJacDot(jnt_array, jac_dot_q_dot);
+      if (error != 0)
+        throw std::runtime_error(solver.strError(error));
+    }
+
+    // Compute the jacbian
+    Eigen::MatrixXd jac = motion_group->calcJacobian(q, tcp);
+
+    // Compute the joint accelerations
+    // d/dt(x_dot) = d/dt(J * q_dot)
+    // x_dot_dot = J_dot * q_dot + J * q_dot_dot
+    // qdotdot = J^(-1) * (x_dot_dot - J_dot_q_dot)
+    Eigen::VectorXd twist = fromKDL(x_dot_dot - jac_dot_q_dot);
+
+    Eigen::Index dof = q.size();
+    Eigen::VectorXd q_dot_dot(dof);
+    if (dof == 6)
+    {
+      q_dot_dot = jac.colPivHouseholderQr().solve(twist);
+    }
+    else
+    {
+      if (!tesseract_kinematics::solvePInv(jac, twist, q_dot_dot))
+        throw std::runtime_error("Failed to solve pseudo-inverse for acceleration calculation");
+    }
+
+    return q_dot_dot;
+  }
+
+  tesseract_kinematics::JointGroup::UPtr motion_group;
+  KDL::Chain kdl_chain_;
+  const std::string tcp;
+  const double max_translational_vel;
+  const double max_translational_acc;
```

Comment:
I feel like these should have an underscore suffix as they are class members. Same with motion_group. It's confusing in the constructor.